book-content/chapters/11-annotations-and-assertions.md

```diff
@@ -0,0 +1,1196 @@
+
+You happen to know that the user can't navigate to this page without passing `?id=some-id` to the URL.
+
+```ts twoslash
+const searchParams = new URLSearchParams(window.location.search);
+
+const id = searchParams.get("id");
+//    ^?
+```
+
+But TypeScript doesn't know that the `id` will always be a string. It thinks that `id` could be a string or `null`.
+
+So, let's force it. We can use `as` on the result of `searchParams.get("id")` to tell TypeScript that we know it will always be a string:
+
+```ts twoslash
+const searchParams = new URLSearchParams(window.location.search);
+// ---cut---
+const id = searchParams.get("id") as string;
+//    ^?
+```
+
+Now TypeScript knows that `id` will always be a string, and we can use it as such.
+
+This `as` is a little unsafe! If `id` is somehow not actually passed in the URL, it will be `null` at runtime but `string` at compile time. This means if we called `.toUpperCase()` on `id`, we'd crash our app.
+
+But it's useful in cases where we truly know more than TypeScript can about the behavior of our code.
+
+#### An Alternative Syntax
+
+As an alternative to `as`, you can prefix the value with the type wrapped in angle brackets:
+
+```typescript
+const id = <string>searchParams.get("id");
+```
+
+This is less common than `as`, but behaves exactly the same way. `as` is more common, so it's better to use that.
+
+#### The Limits of `as`
+
+`as` has some limits on how it can be used. It can't be used to convert between unrelated types.
+
+Consider this example where `as` is used to assert that a string should be treated as a number:
+
+```ts twoslash
+// @errors: 2352
+const albumSales = "Heroes" as number;
+```
+
+TypeScript realizes that even though we're using `as`, we might have made a mistake. The error message is telling us that a string and a number don't share any common properties, but if we really want to go through with it, we could double up on the `as` assertions to first assert the string as `unknown` and then as a `number`:
+
+```tsx
+const albumSales = "Heroes" as unknown as number; // no error
+```
+
+When using `as` to assert as `unknown as number`, the red squiggly line goes away but that doesn't mean the operation is safe. There's just no way to convert `"Heroes"` into a number that would make sense.
+
+The same behavior applies to other types as well.
+
+In this example, an `Album` interface and a `SalesData` interface don't share any common properties:
+
+```ts twoslash
+// @errors: 2352
+interface Album {
+  title: string;
+  artist: string;
+  releaseYear: number;
+}
+
+interface SalesData {
+  sales: number;
+  certification: string;
+}
+
+const paulsBoutique: Album = {
+  title: "Paul's Boutique",
+  artist: "Beastie Boys",
+  releaseYear: 1989,
+};
+
+const paulsBoutiqueSales = paulsBoutique as SalesData;
+```
+
+Again, TypeScript shows us the warning about the lack of common properties.
```

Comment:
since u mention common properties... it's also worth mentioning that TypeScript has a concept of "weak types". In an alternative universe this could be fine but in TS it isn't:
```ts
declare const obj1: { a?: number };

const obj2: { b?: number } = obj1; // error
```

When the type has only optional properties then a common property is still required for the assignment to work.